omnipaxos_core/src/sequence_paxos/follower.rs

```diff
@@ -70,39 +89,80 @@ where
         }
     }
 
+    // Correctness: This function performs multiple operations that cannot be rolled
+    // back, so instead it relies on writing in a "safe" order for correctness.
     pub(crate) fn handle_acceptsync(&mut self, accsync: AcceptSync<T>, from: NodeId) {
-        if self.internal_storage.get_promise() == accsync.n
+        if self.internal_storage.get_promise().expect("storage error") == accsync.n
             && self.state == (Role::Follower, Phase::Prepare)
         {
+            let old_decided_idx = self
+                .internal_storage
+                .get_decided_idx()
+                .expect("storage error");
+            let old_accepted_round = self
+                .internal_storage
+                .get_accepted_round()
+                .expect("storage error");
+            self.internal_storage
+                .set_accepted_round(accsync.n)
+                .expect("storage error");
+            let result = self.internal_storage.set_decided_idx(accsync.decided_idx);
+            if result.is_err() {
+                self.internal_storage
+                    .set_accepted_round(old_accepted_round)
+                    .expect("storage error");
+            }
+            result.expect("storage error");
             let accepted = match accsync.decided_snapshot {
                 Some(s) => {
-                    match s {
+                    let result = match s {
                         SnapshotType::Complete(c) => {
-                            self.internal_storage.set_snapshot(accsync.decided_idx, c);
+                            self.internal_storage.set_snapshot(accsync.decided_idx, c)
                         }
                         SnapshotType::Delta(d) => {
-                            self.internal_storage.merge_snapshot(accsync.decided_idx, d);
+                            self.internal_storage.merge_snapshot(accsync.decided_idx, d)
                         }
+                    };
+                    if result.is_err() {
+                        self.internal_storage
+                            .set_accepted_round(old_accepted_round)
+                            .expect("storage error");
+                        self.internal_storage
+                            .set_decided_idx(old_decided_idx)
+                            .expect("storage error");
                     }
+                    result.expect("storage error");
                     let accepted_idx = self.internal_storage.append_entries(accsync.suffix);
+                    if accepted_idx.is_err() {
+                        // we set the decided snapshot successfully, so no need to rollback decided_idx
```

Comment:
I think since we know the decided snapshot is set (and this includes all decided entries from this AccSync), it makes sense to not rollback the decided index. It's better to leave it consistent with the snapshot/log, right?

Reply:
Right, the decided index and snapshot are consistent. Keep it like this, then.